src/backend/metal/shaders/commands.metal

```diff
@@ -1,67 +1,84 @@
 #include <metal_stdlib>
 using namespace metal;
 
+// -------------- Image Blits -------------- //
+
 typedef struct {
     float4 src_coords [[attribute(0)]];
     float4 dst_coords [[attribute(1)]];
-} TextureBlitAttributes;
+} BlitAttributes;
 
 typedef struct {
     float4 position [[position]];
     float4 uv;
     uint layer [[render_target_array_index]];
-} VertexData;
+} BlitVertexData;
 
-vertex VertexData vs_blit(TextureBlitAttributes in [[stage_in]]) {
+vertex BlitVertexData vs_blit(BlitAttributes in [[stage_in]]) {
     float4 pos = { 0.0, 0.0, 0.0f, 1.0f };
     pos.xy = in.dst_coords.xy * 2.0 - 1.0;
-    return VertexData { pos, in.src_coords, uint(in.dst_coords.z) };
+    return BlitVertexData { pos, in.src_coords, uint(in.dst_coords.z) };
 }
 
 fragment float4 ps_blit_1d(
-    VertexData in [[stage_in]],
+    BlitVertexData in [[stage_in]],
     texture1d<float> tex1D [[ texture(0) ]],
     sampler sampler2D [[ sampler(0) ]]
 ) {
   return tex1D.sample(sampler2D, in.uv.x);
 }
 
 fragment float4 ps_blit_1d_array(
-    VertexData in [[stage_in]],
+    BlitVertexData in [[stage_in]],
     texture1d_array<float> tex1DArray [[ texture(0) ]],
     sampler sampler2D [[ sampler(0) ]]
 ) {
   return tex1DArray.sample(sampler2D, in.uv.x, uint(in.uv.z));
 }
 
 fragment float4 ps_blit_2d(
-    VertexData in [[stage_in]],
+    BlitVertexData in [[stage_in]],
     texture2d<float> tex2D [[ texture(0) ]],
     sampler sampler2D [[ sampler(0) ]]
 ) {
   return tex2D.sample(sampler2D, in.uv.xy, level(in.uv.w));
 }
 
 fragment float4 ps_blit_2d_array(
-    VertexData in [[stage_in]],
+    BlitVertexData in [[stage_in]],
     texture2d_array<float> tex2DArray [[ texture(0) ]],
     sampler sampler2D [[ sampler(0) ]]
 ) {
   return tex2DArray.sample(sampler2D, in.uv.xy, uint(in.uv.z), level(in.uv.w));
 }
 
 fragment float4 ps_blit_3d(
-    VertexData in [[stage_in]],
+    BlitVertexData in [[stage_in]],
     texture3d<float> tex3D [[ texture(0) ]],
     sampler sampler2D [[ sampler(0) ]]
 ) {
   return tex3D.sample(sampler2D, in.uv.xyz, level(in.uv.w));
 }
 
+// -------------- Buffer Fill/Copy -------------- //
+
 kernel void cs_fill_buffer(
     device uint *buffer [[ buffer(0) ]],
     constant uint &value [[ buffer(1) ]],
     uint index [[ thread_position_in_grid ]]
 ) {
     buffer[index] = value;
 }
+
+kernel void cs_copy_buffer(
+    device uint *dest [[ buffer(0) ]],
+    device uint *source [[ buffer(1) ]],
+    constant uint &size [[ buffer(2) ]],
+    uint index [[ thread_position_in_grid ]]
+) {
+    if (index < size) {
+        uint mask = index + 4 <= size ? ~0 : ((1 << (size - index) * 8) - 1);
+        uint wi = index / 4; // word index
+        dest[wi] = (source[wi] & mask) + (dest[wi] & ~mask);
```

Comment:
How do we know `dest[wi]` isn't out-of-bounds?

Reply:
There is a size check two lines above. It can still write up to extra 3 bytes more than the size, but we at least guarantee to not change the contents. I don't know if we can do better here...

Reply:
It seems like undefined behavior if we accidentally write outside the buffer, maybe the value wouldn't be written at all?

Is there a reason we don't want to copy per `uchar` instead of `uint`? If I understand correctly it also removes the need for the mask. We could always optimize this later if we're trying to reduce the amount of writes

Reply:
> It seems like undefined behavior if we accidentally write outside the buffer, maybe the value wouldn't be written at all?

Given that the offset has to be a multiple of 4, we should just round up the buffer sizes to the multiple of 4, which will guarantee that those copies never try to write OOB. I'll add this code.

> Is there a reason we don't want to copy per uchar instead of uint?

I looked it up previously and I don't think this is an option in MSL:
```
Ordinary Metal buffers may contain: 
  - Basic types such as float and int
  - Vector and matrix types
  - Arrays of buffer types
  - Structs of buffer types 
  - Unions of buffer types 
```